src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered <see cref="IHandleLocationChanging"/> handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="forceLoad">Whether the navigation is attempting to bypass client-side routing.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation.</returns>
+    public async ValueTask<bool> NotifyLocationChanging(string uri, bool isNavigationIntercepted, bool forceLoad)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        if (_locationChangingHandlers.Count == 0)
+        {
+            return false;
+        }
+
+        _locationChangingCts = new();
+
+        var cancellationToken = _locationChangingCts.Token;
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, forceLoad, cancellationToken);
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<IHandleLocationChanging>.Shared.Rent(handlerCount);
+        _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+        try
+        {
+            for (var i = 0; i < handlerCount; i++)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
+
+                if (context.IsCanceled || cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
+        {
+            // If it was our cancellation token, we should gracefully cancel the navigation.
+            return true;
+        }
+        catch (Exception ex)
+        {
+            throw new LocationChangeException("An exception occurred while dispatching a location changing event.", ex);
+        }
+        finally
+        {
+            ArrayPool<IHandleLocationChanging>.Shared.Return(locationChangingHandlersCopy);
+        }
+    }
+
+    /// <summary>
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
+    protected virtual bool SetHasLocationChangingHandlers(bool value)
+        => true;
+
+    /// <summary>
+    /// Invokes <see cref="SetHasLocationChangingHandlers(bool)"/> if the existence of location changing handlers has changed.
+    /// </summary>
+    protected void UpdateHasLocationChangingHandlers()
```

Comment:
Is this protected only because of the `AttachJsRuntime` flow for `RemoteNavigationManager`? If so is there any alternative? It seems quite overspecialised to make a protected method for this one thing if it's only relevant in one of the subclasses.

For example, if subclasses could directly read the "navigation locked" status, then `RemoteNavigationManager` could call its own `SetHasLocationChangingHandlers` from `AttachJsRuntime` if the navigation lock is meant to be active, and then `UpdateHasLocationChangingHandlers` could be private.

Or maybe that's not even needed. If `RemoteNavigationManager` receives the `SetHasLocationChangingHandlers` call too early, it could store the flag value and then use it during `AttachJsRuntime`.